all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+# Worker Autotuning Proposal
+
+Users often struggle with the appropriate configuration of workers. There are many options to fiddle with, nearly none
+of which are something users should be required to think about. This proposal aims to reduce the cognitive load of
+configuring workers by automatically tuning various worker internals at runtime without (or with configurable) user
+intervention.
+
+## Areas of configuration
+There are a few different configuration areas that we can autotune:
+
+### Concurrent Pollers
+Right now the number of pollers used for polling workflow and activity tasks is fixed to a maximum defined by the user.
+In Core: `max_concurrent_wft_polls` and `max_concurrent_at_polls`. There's no need for this number to be fixed, and we
+can and should scale it up and down as demand dictates.  This work is largely already designed and implemented in Go:
+[see the branch
+here](https://github.com/temporalio/sdk-go/compare/master...Quinn-With-Two-Ns:sdk-go:poller-autottune-v0.2). Hence this
+doc will not focus on this area in detail.
+
+It is important to note that there is a distinction between the active number of poller *routines* and the active number
+of *open actual RPC calls*. The number of routines may be scaled independently of the number of open RPC calls at any
+time. The former may have a floor, whereas the latter must always be able to go to 0. Users are unlikely to care about
+the number of routines as they are relatively cheap. They *do*, however, care about how quickly we will change the
+number of active RPCs. The floor on the number of routines effectively determines how quickly we are willing to consume
+available slots and thus make actual RPC calls. IE: If the floor on the number of routines is 10, and 10 slots open up
+at the same time, we will immediately make 10 RPC calls. 
+
+In Go the name for these options is already misleading: `MaxConcurrentWorkflowTaskPollers` is not accurate, it's really
+`NumWorkflowTaskPollers` (or routines), what it probably should be is `MaxConcurrentWorkflowTaskPolls`, since the number
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
```

Comment:
It might be nice to say *something* here about what this might look like.  For instance, we could take the approach that we tune for maximum concurrency (maintaining OOM or other error avoidance), or total throughput per worker, or minimum latency (for some measure of latency and some form of "minimum").  Any (or even even each) of these seem like valid strategies, as might some utility function that combined them.  Do we plan to offer the user a choice?  Pick one of these (or something I didn't mention) and do just that for now?  What's our true goal here?

Even though I have placed my comment here in the slot/cache tuning section, it's really more of a global comment.  We can probably do some stuff that improves worker performance in ways that are pareto optimal ... i.e. improve some dimension of performance without any cost in other dimensions.  Maybe that should be our first goal.  But most (I might even say all) performance tuning eventually requires a definition of performance that expresses the specific goal directly because there are always tradeoffs.

Reply:
Note that the "default implementations" section does make some reference to those.